audittools/backing_store_sql.go

```diff
@@ -0,0 +1,313 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company
+// SPDX-License-Identifier: Apache-2.0
+
+package audittools
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+
+	"github.com/lib/pq"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-api-declarations/cadf"
+)
+
+var sqlIdentifierRegex = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*$`)
+
+func init() {
+	// Self-register with the backing store registry
+	BackingStoreRegistry.Add(func() BackingStore {
+		return &SQLBackingStore{}
+	})
+}
+
+// SQLBackingStore implements BackingStore using a PostgreSQL database.
+//
+// Thread safety: All operations use database transactions for atomicity.
+// Multiple concurrent calls are safe and will be serialized by the database.
+// This implementation is suitable for services that already have a database
+// connection and want to avoid managing filesystem volumes.
+//
+// Database Schema:
+// The backing store requires a table with the following schema:
+//
+//	CREATE TABLE audit_events (
+//	    id BIGSERIAL PRIMARY KEY,
+//	    event_data JSONB NOT NULL,
+//	    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
+//	);
+//	CREATE INDEX ON audit_events (created_at, id);
+type SQLBackingStore struct {
+	// Configuration (JSON params)
+	DSN           string `json:"dsn"`            // Database connection string
+	TableName     string `json:"table_name"`     // Table name (default: "audit_events")
+	BatchSize     int    `json:"batch_size"`     // Number of events per batch (default: 100)
+	MaxEvents     int    `json:"max_events"`     // Maximum total events to buffer (default: 10000)
+	DriverName    string `json:"driver_name"`    // SQL driver name (default: "postgres")
+	SkipMigration bool   `json:"skip_migration"` // Skip automatic table creation (default: false)
+
+	// Runtime state (not from JSON)
+	db           *sql.DB                `json:"-"`
+	writeCounter prometheus.Counter     `json:"-"`
+	readCounter  prometheus.Counter     `json:"-"`
+	errorCounter *prometheus.CounterVec `json:"-"`
+	sizeGauge    prometheus.Gauge       `json:"-"`
+}
+
+// PluginTypeID implements pluggable.Plugin.
+func (s *SQLBackingStore) PluginTypeID() string {
+	return "sql"
+}
+
+// Init implements BackingStore.
+func (s *SQLBackingStore) Init(registry prometheus.Registerer) error {
+	// Validate configuration
+	if s.DSN == "" {
+		return errors.New("audittools: dsn is required for sql backing store")
+	}
+
+	// Set defaults
+	if s.DriverName == "" {
+		s.DriverName = "postgres"
+	}
+	if s.TableName == "" {
+		s.TableName = "audit_events"
+	}
+	if s.BatchSize == 0 {
+		s.BatchSize = 100
+	}
+	if s.MaxEvents == 0 {
+		s.MaxEvents = 10000
+	}
```

Comment:
For optional values with defaults, consider declaring the field as [`Option[]`](https://pkg.go.dev/github.com/majewsky/gg/option#Option) and then using it as `s.TableName.UnwrapOr("audit_events")` etc.